### PR TITLE
Remove `petgraph` from logical plan dependency

### LINKED ISSUE #206: Remove Petgraph dependency from `LogicalGraph`
As a follow on action item from #202  this commit removes the `petgraph`
dependency from `LogicalPlan`.

### DoD
- The dependency is removed from `partiql-logical` and the current tests pass.

### PR DESCRIPTION
**Pull request by am357**
*Issue #, if available:* 

Closes #206

*Description of changes:*
    
As a follow on action item from #202  this commit removes the `petgraph`
dependency from `LogicalPlan`.
    
The graph presented in this commit has just enough interfaces for doing
he work for now. We need to iterate on it later when vending as an API
to consumers.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-logical/src/lib.rs [hunk 1]
```diff
@@ -1,14 +1,58 @@
...
+#[derive(Debug, Hash)]
+pub struct OpId(usize);
+
+impl OpId {
+    pub fn index(&self) -> usize {
+        self.0
+    }
+}
+
+impl PartialEq<Self> for OpId {
```
Comment (jpschorr) [inline partiql-logical/src/lib.rs:13]: can probably use `#[derive(PartialEq, Eq)]`

File: partiql-logical/src/lib.rs [hunk 2]
```diff
@@ -1,14 +1,58 @@
 use partiql_value::{BindingsName, Value};
-use petgraph::prelude::StableGraph;
-use petgraph::Directed;
 use std::collections::HashMap;
 
+#[derive(Debug, Hash)]
```
Comment (jpschorr) [inline partiql-logical/src/lib.rs:4]: for wrapping a `usize`, you could also `#derive[(Copy)]` here, which would make dealing with these IDs more pleasant

File: partiql-logical/src/lib.rs [hunk 3]
```diff
@@ -1,14 +1,58 @@
...
+}
+
+impl Eq for OpId {}
+
 #[derive(Debug)]
-pub struct LogicalPlan(pub StableGraph<BindingsExpr, (), Directed>);
+pub struct LogicalPlan<'a, T> {
+    nodes: Vec<T>,
+    edges: Vec<(&'a OpId, &'a OpId)>,
+    node_count: usize,
```
Comment (jpschorr) [inline partiql-logical/src/lib.rs:25]: is this not redundant with `nodes.len()` ?

File: partiql-logical/src/lib.rs [hunk 4]
```diff
@@ -1,14 +1,58 @@
...
+    }
+}
+
+impl Eq for OpId {}
+
 #[derive(Debug)]
-pub struct LogicalPlan(pub StableGraph<BindingsExpr, (), Directed>);
+pub struct LogicalPlan<'a, T> {
+    nodes: Vec<T>,
+    edges: Vec<(&'a OpId, &'a OpId)>,
```
Comment (jpschorr) [inline partiql-logical/src/lib.rs:24]: I don't think it's worth the hassle of references for the ids here. Might as well use `Vec<(OpId, OpId)>` and ensure the invariant that the ids must reference existing nodes in the `add_flow` fn.

File: partiql-logical/src/lib.rs (hunk 3 above)
Comment (jpschorr) [reply partiql-logical/src/lib.rs:25]: e.g. 


```rust

pub fn node_count(&self) -> usize {self.nodes.len()}

```

File: partiql-logical/src/lib.rs [hunk 5]
```diff
@@ -1,14 +1,58 @@
...
+    pub fn add_operator(&mut self, op: T) -> OpId {
+        self.nodes.push(op);
+        self.node_count += 1;
+        OpId(self.node_count)
+    }
+
+    pub fn add_flow<'a: 'b>(&mut self, src: &'a OpId, dst: &'a OpId) {
+        let src_index = src.index() - 1;
+        let dst_index = dst.index() - 1;
+        self.nodes.get(src_index).expect("No such operator exists");
```
Comment (jpschorr) [inline partiql-logical/src/lib.rs:46]: this is essentially 
```rust

assert!(src <= self.node_count())

```

Comment (am357) [reply partiql-logical/src/lib.rs:46]: True, will change.

File: partiql-logical/src/lib.rs (hunk 3 above)
Comment (am357) [reply partiql-logical/src/lib.rs:25]: It is actually redundant, will change.

File: partiql-logical/src/lib.rs (hunk 1 above)
Comment (am357) [reply partiql-logical/src/lib.rs:13]: Yeah I think what confused me was a compiler error that I got during dev; let me take another stab at it.

File: partiql-logical/src/lib.rs (hunk 4 above)
Comment (am357) [reply partiql-logical/src/lib.rs]: Makes sense; changed in the new commit.

File: partiql-logical/src/lib.rs [hunk 6]
```diff
@@ -1,14 +1,53 @@
...
+
+    pub fn add_operator(&mut self, op: T) -> OpId {
+        self.nodes.push(op);
+        OpId(self.operator_count())
+    }
+
+    pub fn add_flow(&mut self, src: OpId, dst: OpId) {
+        let src_idx = src.index() - 1;
+        let dst_idx = dst.index() - 1;
+        assert!(src_idx <= self.operator_count());
```
Comment (jpschorr) [inline partiql-logical/src/lib.rs:35]: I believe
- created `OpId`s are 1-indexed (they are created after the new node is `push`ed). 
- `operator_count()` is also '1-indexed' 

so either the `-1` is unnecessary or we should replace `<=` with `<`

**Review by jpschorr — APPROVED** [review]

File: partiql-logical/src/lib.rs (hunk 6 above)
Comment (am357) [reply partiql-logical/src/lib.rs:35]: Agreed, I overlooked this as I was rushing to push the commit; will change in the PR that I'm putting together for moving the rest of the operators to DAG plan.

**Pull request merged**